File: android/app/src/main/java/com/lunacoding/naiclova/view/KeywordView.java
package com.lunacoding.naiclova.view;

import android.content.Context;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.lunacoding.naiclova.R;

public class KeywordView extends RelativeLayout {
	Context context;
	
	RelativeLayout RL_root;
	LinearLayout LL_background;
	TextView TV_textview;
	
	boolean isChecked = false;
	
	private void initView(Context context) {
		this.context = context;
		
		String infService = Context.LAYOUT_INFLATER_SERVICE;
		LayoutInflater li = (LayoutInflater) getContext().getSystemService(infService);
		View v = li.inflate(R.layout.keyword, this, false);
		addView(v);
		
		// set view
		RL_root = findViewById(R.id.root);
		LL_background = findViewById(R.id.background);
		TV_textview = findViewById(R.id.textview);
		
		// event
		RL_root.setOnClickListener(new OnClickListener() {
			@Override
			public void onClick(View view) {
				onViewClick();
			}
		});
		RL_root.setOnLongClickListener(new OnLongClickListener() {
			@Override
			public boolean onLongClick(View view) {
				onViewLongClick();
				return false;
			}
		});
	}
	
	public KeywordView(Context context) {
		super(context);
		initView(context);
	}
	
	public KeywordView(Context context, AttributeSet attrs) {
		super(context, attrs);
		initView(context);
	}
	
	public KeywordView(Context context, AttributeSet attrs, int defStyleAttr) {
		super(context, attrs, defStyleAttr);
		initView(context);
	}
	
	public KeywordView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
		super(context, attrs, defStyleAttr, defStyleRes);
		initView(context);
	}
	
	public void setChecked(boolean b) {
		isChecked = b;
		
		LL_background.setBackground(getResources().getDrawable(b ? R.drawable.word_on : R.drawable.word_off));
		TV_textview.setTextColor(b ? 0xffaaaaaa : 0xffffffff);
	}
	
	public boolean isChecked() {
		return isChecked;
	}
	
	public void setText(String msg) {
		TV_textview.setText(msg);
	}
	
	public String getText(){
		return TV_textview.getText().toString();
	}
	
	// =========================================================================================
	
	private OnEventListener onEventListener = null;
	
	public interface OnEventListener {
		
		void onViewClick(KeywordView v);
		
		void onViewLongClick(KeywordView v);
	}
	
	public KeywordView setOnEventListener(OnEventListener listener) {
		this.onEventListener = listener;
		return this;
	}
	
	public void onViewClick() {
		if (onEventListener != null) onEventListener.onViewClick(this);
	}
	
	public void onViewLongClick() {
		if (onEventListener != null) onEventListener.onViewLongClick(this);
	}
}

File: android/app/src/main/java/com/lunacoding/naiclova/view/KeywordView.kt
package com.lunacoding.naiclova.view

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View.OnLongClickListener
import android.widget.LinearLayout
import android.widget.RelativeLayout
import android.widget.TextView
import com.lunacoding.naiclova.R

class KeywordView : RelativeLayout {
	var RL_root: RelativeLayout? = null
	var LL_background: LinearLayout? = null
	var TV_textview: TextView? = null
	var isChecked = false
		set(value) {
			field = value
			LL_background!!.background =
					resources.getDrawable(if (value) R.drawable.word_on else R.drawable.word_off)
			TV_textview!!.setTextColor(if (value) -0x555556 else -0x1)
		}

	private fun initView() {
		val infService = Context.LAYOUT_INFLATER_SERVICE
		val li = getContext().getSystemService(infService) as LayoutInflater
		val v = li.inflate(R.layout.view_keyword, this, false)
		addView(v)

		// set view
		RL_root = findViewById(R.id.root)
		LL_background = findViewById(R.id.background)
		TV_textview = findViewById(R.id.textview)

		// event


		RL_root?.setOnClickListener {
			onViewClick()
		}
		RL_root?.setOnLongClickListener {
			onViewLongClick()
			false
		}
	}

	constructor(context: Context) : super(context) {
		initView()
	}

	constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
		initView()
	}

	constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
			context,
			attrs,
			defStyleAttr
	) {
		initView()
	}

	constructor(
			context: Context,
			attrs: AttributeSet?,
			defStyleAttr: Int,
			defStyleRes: Int
	) : super(context, attrs, defStyleAttr, defStyleRes) {
		initView()
	}

	var text: String?
		get() = TV_textview!!.text.toString()
		set(msg) {
			TV_textview!!.text = msg
		}

	// =========================================================================================
	private var onEventListener: OnEventListener? = null

	interface OnEventListener {
		fun onViewClick(v: KeywordView)
		fun onViewLongClick(v: KeywordView?)
	}

	fun setOnEventListener(listener: OnEventListener?): KeywordView {
		onEventListener = listener
		return this
	}

	fun onViewClick() {
		if (onEventListener != null) onEventListener!!.onViewClick(this)
	}

	fun onViewLongClick() {
		if (onEventListener != null) onEventListener!!.onViewLongClick(this)
	}
}